src/workerd/server/workerd.c++

```diff
@@ -486,11 +487,125 @@ kj::Maybe<kj::Own<capnp::SchemaFile>> tryImportBulitin(kj::StringPtr name) {
 
 // =======================================================================================
 
+// A kj::Network implementation which wraps some other network and optionally (if enabled)
+// implements "loopback:" network addresses, which are expected to be serviced within the same
+// process. Loopback addresses are enabled only when running `workerd test`. The purpose is to
+// allow end-to-end testing of the network stack without creating a real external-facing socket.
+//
+// There is no use for loopback sockets in production since direct service bindings are more
+// efficient while solving the same problems.
+class NetworkWithLoopback final: public kj::Network {
+public:
+  NetworkWithLoopback(kj::Network& inner, kj::AsyncIoProvider& ioProvider)
+      : inner(inner), ioProvider(ioProvider), loopbackEnabled(rootLoopbackEnabled) {}
+
+  NetworkWithLoopback(kj::Own<kj::Network> inner, kj::AsyncIoProvider& ioProvider,
+                      bool& loopbackEnabled)
+      : inner(*inner), ownInner(kj::mv(inner)), ioProvider(ioProvider),
+        loopbackEnabled(loopbackEnabled) {}
+
+  // Call once to enable loopback addresses.
+  void enableLoopback() { loopbackEnabled = true; }
+
+  kj::Promise<kj::Own<kj::NetworkAddress>> parseAddress(
+      kj::StringPtr addr, uint portHint = 0) override {
+    if (loopbackEnabled && addr.startsWith(PREFIX)) {
+      return kj::Own<kj::NetworkAddress>(kj::heap<LoopbackAddr>(*this, addr.slice(PREFIX.size())));
+    } else {
+      return inner.parseAddress(addr, portHint);
+    }
+  }
+
+  kj::Own<kj::NetworkAddress> getSockaddr(const void* sockaddr, uint len) override {
+    return inner.getSockaddr(sockaddr, len);
+  }
+
+  kj::Own<kj::Network> restrictPeers(
+      kj::ArrayPtr<const kj::StringPtr> allow,
+      kj::ArrayPtr<const kj::StringPtr> deny = nullptr) override {
+    return kj::heap<NetworkWithLoopback>(
+        inner.restrictPeers(allow, deny), ioProvider, loopbackEnabled);
+  }
+
+private:
+  kj::Network& inner;
+  kj::Own<kj::Network> ownInner;
+  kj::AsyncIoProvider& ioProvider KJ_UNUSED;
+  bool rootLoopbackEnabled = false;
+  bool& loopbackEnabled;
```

Comment:
A brief comment indicating who owns this `bool&` would be helpful.

Reply:
6719d3db